go.mod

```diff
@@ -1,100 +1,50 @@
-	github.com/openshift/api v0.0.0-20180830153656-5ad8479f64f1 // indirect
-	github.com/openshift/client-go v0.0.0-20180830153425-431ec9a26e50 // indirect
-	github.com/pborman/uuid v0.0.0-20150603214016-ca53cad383ca // indirect
-	github.com/peterbourgon/diskv v2.0.1+incompatible // indirect
-	github.com/pkg/errors v0.8.1 // indirect
-	github.com/prometheus/client_golang v0.0.0-20170531130054-e7e903064f5e // indirect
-	github.com/prometheus/client_model v0.0.0-20150212101744-fa8ad6fec335 // indirect
-	github.com/prometheus/common v0.0.0-20170427095455-13ba4ddd0caa // indirect
-	github.com/prometheus/procfs v0.0.0-20170519190837-65c1f6f8f0fc // indirect
-	github.com/sirupsen/logrus v0.0.0-20170822132746-89742aefa4b2 // indirect
-	github.com/spf13/afero v1.2.2 // indirect
-	github.com/spf13/cobra v0.0.0-20180319062004-c439c4fa0937
-	github.com/spf13/pflag v1.0.1
-	github.com/stretchr/testify v1.3.0 // indirect
-	github.com/ugorji/go v0.0.0-20170107133203-ded73eae5db7 // indirect
+	github.com/spf13/cobra v0.0.5
+	github.com/spf13/pflag v1.0.5
 	github.com/xiang90/probing v0.0.0-20190116061207-43a291ad63a2 // indirect
-	go.etcd.io/bbolt v1.3.3 // indirect
-	golang.org/x/crypto v0.0.0-20180808211826-de0752318171 // indirect
-	golang.org/x/sync v0.0.0-20190423024810-112230192c58 // indirect
-	golang.org/x/time v0.0.0-20161028155119-f51c12702a4d // indirect
-	google.golang.org/appengine v1.4.0 // indirect
-	gopkg.in/airbrake/gobrake.v2 v2.0.9 // indirect
-	gopkg.in/gemnasium/logrus-airbrake-hook.v2 v2.1.2 // indirect
-	gopkg.in/inf.v0 v0.9.0 // indirect
-	gopkg.in/natefinch/lumberjack.v2 v2.0.0-20150622162204-20b71e5b60d7 // indirect
-	gopkg.in/square/go-jose.v2 v2.0.0-20180411045311-89060dee6a84 // indirect
-	gopkg.in/yaml.v1 v1.0.0-20140924161607-9f9df34309c0 // indirect
-	gopkg.in/yaml.v2 v2.2.1
-	gotest.tools v2.2.0+incompatible // indirect
-	k8s.io/api v0.0.0-20181128191700-6db15a15d2d3
-	k8s.io/apiextensions-apiserver v0.0.0-20181217193801-853f76028711 // indirect
-	k8s.io/apimachinery v0.0.0-20181128191346-49ce2735e507
-	k8s.io/apiserver v0.0.0-20181207191401-9601a7bf41ef
-	k8s.io/client-go v0.0.0-20181205110444-5e6a3d4e34f6
-	k8s.io/csi-api v0.0.0-20181121073239-b5848c3064d4 // indirect
-	k8s.io/kube-controller-manager v0.0.0-20181110193944-1c4e86102515 // indirect
-	k8s.io/kube-openapi v0.0.0-20180711000925-0cf8f7e6ed1d // indirect
-	k8s.io/kubernetes v1.12.4
-	k8s.io/utils v0.0.0-20180726175726-66066c83e385 // indirect
-	kubevirt.io/containerized-data-importer v1.5.1 // indirect
-	kubevirt.io/kubevirt v0.16.0
+	gopkg.in/yaml.v2 v2.2.8
+	k8s.io/api v0.17.0
+	k8s.io/apimachinery v0.17.1-beta.0
+	k8s.io/apiserver v0.16.8
+	k8s.io/client-go v12.0.0+incompatible
+	k8s.io/cloud-provider v0.16.8
+	k8s.io/component-base v0.17.0
+	k8s.io/klog v1.0.0
+	k8s.io/kubernetes v1.16.8
+	k8s.io/utils v0.0.0-20191114184206-e782cd3c129f // indirect
+	kubevirt.io/client-go v0.26.5
+)
+
+replace (
```

Comment:
Why do we need so many "replace" dependencies? I would prefer to have most of the dependencies as normal "require" dependencies and only use "replace" when we need to overwrite something.

Reply:
It's a "best practice" from other CCMs, see for example https://github.com/kubernetes/cloud-provider-gcp/blob/master/go.mod, or our fork of the same https://github.com/gardener/cloud-provider-gcp/blob/release-v1.17/go.mod, and if you search for other CCMs you will find it's a pattern. The rationale is that CCM is actually including a lot of Kubernetes core code, and we would like to make sure that all dependencies come from "the same" Kubernetes to avoid any unforeseen issues. This is hard to guarantee with Go modules unless explicit replace dependencies are used.

I could probably remove the replace dependencies and it might even work, but it would probably be more fragile and prone to issues in the future.